26-dracula-ant/technical_specification.md

```diff
@@ -55,38 +57,59 @@ Deletion of work packages should be logged or implemented as deactivation (immed
+
+- `GET /download-access/users/{user_id}/datasets/{dataset_id}`
   - authorization: only internal from download controller (via Istio)
   - returns `true` or `false` as a scalar resource
 
-In order to facilitate authorization, the path of this endpoint starts with `datasets` and not with `users` which is already used by other endpoints fo the claims repository.
+*TODO:* Instead of true/false, the latter endpoint could also return null or the date when the access expires (null could be confused with unlimited access though). This could be used by the Work Package Service to limit the expiration date of the tokens, even though the Work Package Service re-checks access on every operation anyway.
+
+In order to facilitate authorization, the path of these endpoint starts with `download-access` and not with `users` which is already used by other endpoints of the user registry and claims repository.
+
+Note that this is a shortcut as long as we don't have a visa issuer service. Later, the claims repository should not be contacted directly, but the information about access grants will be requested from the visa issuer service and checked with the help of the visa library.
+
+In order to check whether a given user has *upload* access to a given dataset, we may introduce a similar endpoint that will be based on checking custom visa types for dataset submission in the claims repository. Alternatively, we could store a list of internal user IDs of dataset submitters also in the `Dataset` collection described below and update them similarly to the file IDs belonging to the dataset.
 
-This information will later be provided by the visa issuer service and checked with the help of the visa library.
+*TODO:* This can be decided later, since for now only download will be implemented.
 
-In order to check whether a given user has *upload* access to a given dataset, we may introduce a similar endpoint that will be based on checking custom visa types for dataset submission in the claims repository. Alternatively, we could store a list of internal user IDs of dataset submitters also in the `DatasetFiles` collection described below and update them similarly to the file IDs belonging to the dataset. (To be clarified, will be implemented later.)
+#### Access checks by the Frontend
+
+In order to provide the user with a list of downloadable datasets, the Work Package Service combines the information from the access checks explained above with information about the dataset as explained below and makes it avaliable to the frontend via the following endpoint:
+
+- `GET /datasets`
+  - auth header: internal access token with the user context
+  - returns a list of all dataset IDs that can be downloaded by the user
+
+*TODO*: Maybe this should be `/users/{user_id}/datasets` to make it more RESTful? However, the user registry has the same prefix, and the API gateway can only use the prefix to map requests to services. Any better idea?
```

Comment:
Yes that sounds good.

Reply:
Each service should have a unique prefix either way

Reply:
> Yes that sounds good.

If our public API has the prefix `api` and the work package service has its own additional prefix `wps`, then the URL for getting work-packages would be `/api/wps/work-packages`. I don't really like this since it contains duplication and makes the implementation details (services) publicly visible instead of providing an implementation-independent API - `/api/work-packages` would be nicer.

Reply:
We don't have one single integrated API landscape at the moment and we didn't plan to do so. The prefixes can be seen as namespaces or partitions of the API. This allows us to think of the same resources in different contexts in a different way. This is not an uncommon practice, the GA4GH is promoting a similar pattern. E.g. for getting a file via DRS you need to contact https://drs.example.org/ga4gh/drs/v1/objects/{object_id}. There even the version of the service API is visible. If you like, we can put onto the refactoring list to discuss the pros and cons of that approach. However, for now, we can use the flexibility gained by the namespacing. Thus multiple services can use the same resources in their endpoints.